### DISCUSSION TITLE
Performance

### DISCUSSION THREAD

**Discussion by paulobreim** [sepinf-inc/IPED#474] [Q&A]
I would like to know who is using IPED effectively, that is, to work on an expert investigation.
I'm indexing 9 hard drives of 1TB each, using a computer with:
Xeon E5-2658A
64GB Ram
SSD 1.5TB
To index the 1TB HD it takes from 16 hours to 30 hours.
I'm using a forensic profile.

I don't know how to analyze why it is taking so long. Would anyone know how to say?
Thank you,
Paulo

**Comment by wladimirleite** [comment]
Can you share a log file? 
It doesn't need to be the full file, I would like to take a look at its beginning/end. Something like the first and the last 1000 lines should be enough.

**Comment by lfcnassif** [comment]
Check if you are following the updated performance tips: https://github.com/sepinf-inc/IPED/wiki/Performance-Tips

**Reply by lfcnassif** [reply]
updated again.

**Reply by paulobreim** [reply]
Here are 2 good examples to analyze
There are 2 logs on 2 different computers processing a 1TB hard drive.
The two are still running, the first running at 30 hours and 23 minutes and reporting that there are still 78 hours to go, the second is running at 21 hours and 43 minutes and there are still 60 hours and 22 minutes to go.
From what I observe , it's that .xls .xlsb files seem to take a long time.
There are no errors in the command line


I have anydesk installed on both computers, so if you want to connect to see it, just let me know that I pass the password by email. These computers are dedicated to this task, so there are no problems.


Tks
[ipedlog.zip](https://github.com/sepinf-inc/IPED/files/6330252/ipedlog.zip)

**Reply by wladimirleite** [reply]
I will take a closer look at the logs as soon as possible and will reply here.
As the end of the log contains useful information (like time spent in each task), please share it again when (if) the processing is done.
Meanwhile, a screen shot of IPED's window would be useful.

**Reply by paulobreim** [reply]
Sure. Here is 4 images 

sorrry: image was with sensitify datas

**Reply by lfcnassif** [reply]
Looking at the pictures, seems to be an important bottleneck in Tika's OOXmlParser (responsible for xlsx and other office files). Try to use the last master branch build where Tika was upgraded from 1.20 to 1.26, maybe they improved that. You must be logged to download: https://github.com/sepinf-inc/IPED/actions/runs/749584331

**Reply by paulobreim** [reply]
good. I will stop one of than and restart with this version and --continue
thank you

**Reply by lfcnassif** [reply]
This probably will not be possible. 4.0.0-snapshot is not compatible with previous versions. You will have to restart from scratch to test.

**Reply by paulobreim** [reply]
ok. I will

**Reply by paulobreim** [reply]
I think the problem is the same, but if I have this problem everyone has the same problem. It can be tested with the previous version of we are sure that it did not have this problem. 3.17 3.16...


tks

**Reply by paulobreim** [reply]
Well I am restarted the test with IPED 3.16.3 and after 30 minutes, the situation is the same. All tasks processing xlsx, xlsb spending a lot of time for each file. Really very strange.

**Reply by lfcnassif** [reply]
Not so strange, you have very big xlsx files to process, and it is not a simple format like one could imagine, so we use Tika and POI libraries to process them. I see lots of OOME errors in your logs (and hundreds of timeouts), because you enabled enabledExternalParsing (as recommended for OOME errors) but kept the default externalParsingMaxMem = 512M and that is not enough to process your big xlsx files, so lots of OOME errors are being thrown, and before them processing becomes very slow when GC tries to reclaim garbage memory. I also saw in one of your processing you limited the memory to -Xmx8G.

So, I recommend trying 4.0.0-snapshot/alpha again (where some OOM issues - not sure if the same of yours - and BitSet memory usage was improved in POI: https://poi.apache.org/changes.html), **disable enabledExternalParsing** in your forensic profile, and run the processing with java -Xms30G -Xmx30G in your 64GB RAM computer, maybe that will workaround the problems.

**Reply by lfcnassif** [reply]
If you still have problems, please take a thread and a heap dump as described at https://github.com/sepinf-inc/IPED/wiki/Troubleshooting so we could try to investigate and maybe contact POI project to try to fix. Some of your xlsx samples would help a lot.

**Reply by lfcnassif** [reply]
These possibly are related:
https://bz.apache.org/bugzilla/show_bug.cgi?id=64015
https://bz.apache.org/bugzilla/show_bug.cgi?id=63940

**Reply by wladimirleite** [reply]
> Not so strange, you have very big xlsx files to process, and it is not a simple format like one could imagine, so we use Tika and POI libraries to process them. I see lots of OOME errors in your logs, because you enabled enabledExternalParsing (as recommended for OOME errors) but kept the default externalParsingMaxMem = 512M and that is not enough to process your big xlsx files, so lots of OOME errors are being thrown, and before them processing becomes very slow when GC tries to reclaim garbage memory. I also saw in one of your processing you limited the memory to -Xmx8G.
> 
> So, I recommend trying 4.0.0-snapshot/alpha again (where some OOM issues - not sure if the same of yours - and BitSet memory usage was improved in POI: https://poi.apache.org/changes.html), **disable enabledExternalParsing** in your forensic profile, and run the processing with java -Xms30G -Xmx30G in your 64GB RAM computer, maybe that will workaround the problems.

I just ran a local test with ~300K MS Excel files collected from hundreds of cases (~100GB of data), and the parsing took a "reasonable" time. 
Even the largest files (100-200 MB) took ~1-2 minutes to be processed, using a similar configuration that @lfcnassif suggested (IPED 4.0, enabledExternalParsing=false and 30GB for IPED process).

It is quite possible though that peculiar XLSX file can take much longer to be processed, which may be the case of the files you are dealing with.
I tried to manually create some huge files, but they all ended up being handled well by POI -> Tika -> IPED.

@paulobreim, another suggestion would be to export some of these large MS Excel files and run tests only on a few of them (even just one) and see how it goes, in order to isolate the problem, instead of processing the whole evidence file to test new configurations.

As @lfcnassif said, if these some of these files can be sent privately, it would be great.

**Reply by paulobreim** [reply]
Good idea. Now I am runnig the test as @lfcnassif told.

Another test that I would like to do is ignire xlsx files.
Are there a way to setuo a option to ignore xlsx files? 

ps: opss I removed the images because they had sensitive data

**Reply by lfcnassif** [reply]
Great @tc-wleite! Sao Paulo case database is really awesome, thanks for testing!

@paulobreim, yes, you can ignore specific files, as explained in the "processing frozen" section of the troubleshooting link I posted before.

**Reply by lfcnassif** [reply]
Just reviewed Tika's OfficeParserConfig options and they have an "includeShapeBasedContent" option defaulted to true, maybe disabling that could help. Go to conf/ParserConfig.xml and try adding the following parameter to OOXMLParser config block:

&lt;param name="includeShapeBasedContent" type="bool"&gt;false&lt;/param&gt;

**Reply by paulobreim** [reply]
Good! Since I started with your suggestion using java -Xms30G -Xmx30G and disable enabledExternalParsing , look the diference compared with picture that I sent before.

![image](https://user-images.githubusercontent.com/381594/115166852-9ffd2d80-a08b-11eb-9da0-2bd3b05d2548.png)

Practically finished, but now it is just 1 task parsing a trashfile .xlsx with 224MB.
Apparently stopped because the end time no longer updates.
The log is giant.

I will wait by 1 hour ans if not finished I will cancel and retar with --continue.

tks

**Reply by lfcnassif** [reply]
Great! Stopping and using --continue probably will not work, the xlsx file processing will be retried, unless you make an explicit configuration to ignore the file as I pointed. Possibly it is better to wait the timeout to be triggered.

Providing this and other xlsx files would help us trying to improve Tika/POI xlsx parsing...

**Reply by paulobreim** [reply]
uauuu logfile has 5GB.
I will get it.

And I will run again with <param name="includeShapeBasedContent" type="bool">false</param>

And get some .xlsx files.

**Reply by paulobreim** [reply]
@tc-elite really the problem is when we have large xls files. I sent the log and 10 xlsx files do @lfcnassif and now we have examples to analise the problem.

**Comment by paulobreim** [comment]
New Tests.
I finished out 9 complete tests, all with HDs of real cases, 8  which  1TB and 1 witch 512GB, using the IPED version 4 with forensic profile.
I used 2 computers 1 with 64GB and the other with 16GB. It was only possible to finish all of them on the 64GB computer with the suggested java parameters.
 I will do new tests with the next version.

**Reply by lfcnassif** [reply]
You were possibly affected by #539
